\chapter{Background}
\label{ch:background}



\graphicspath{{snoga/}}
%\section{Computational Phylogenetics}
Computational phylogenetics deals with the design and application of computational tools and techniques to conduct necessary analyses related to phylogeny estimation. The primary aim is to reconstruct a most probable hypothesis about the evolutionary ancestry of a set of taxa (i.e., genes, species, etc.). Here we briefly describe the preliminary concepts involved in computational phylogenetics. 
%\begin{comment}
\section{Phylogeny}
Phylogeny refers to the evolutionary relationships among a set of entities such as species, genes, etc popularly represented as a tree known as phylogenetic tree. The entities under consideration are generally referred to as taxa. Each such taxon is placed as a leaf node in a phylogenetic tree and the tree topology represents the evolutionary history. The  internal nodes of the tree represent hypothetical ancestral taxa from which the present day taxa evolved. These ancestral taxa are believed to have existed in the past, but has become extinct. Notably, a tree $T$ is a connected acyclic graph with a set of vertices $V$ and a set of edges $E$. An edge $e = (u, v) \in E$ represents an evolutionary relationship between the two taxa represented by the vertices $u$ and $v$. 

\begin{figure}[!htbp]
\centering
\includegraphics[width=0.9\textwidth]{Figure/outgroup.eps}
\caption{A phylogenetic tree relating four species: human, chimpanzee, gorilla and orangutan. }
\label{fig:outgroup}
\end{figure}

Fig.~\ref{fig:outgroup} shows an example phylogenetic tree among four species: human, chimpanzee, gorilla and orangutan. This evolutionary tree depicts that human and chimpanzee share a common ancestor. As such, we consider humans to be more closely related to chimpanzees than they are to gorillas and orangutans.
%\end{comment}
%\subsubsection{Rooted and unrooted trees} 
In a rooted phylogenetic tree (left part of Fig.~\ref{fig:outgroup}), one vertex $r \in V$ is designated as the root of the tree whereas there is no such designated vertex in an unrooted tree (right part of Fig.~\ref{fig:outgroup}). True evolutionary histories are better represented by a rooted tree. However, identifying the root of an estimated phylogenetic tree is generally a difficult task.

\subsection{Phylogeny Estimation}

\subsection{Evaluation of Phylogeny Estimation Methods}\label{sec:phyPerf}
The performance of a species tree estimation is evaluated by comparing its output tree to the ground truth (i.e., true species tree provided with the simulated dataset) using False Negative (FN) rate also known as the missing branch rate. FN rate actually expresses (in percentage) the fraction of edges that exist in the latter (i.e., in true tree) but are absent in the former (i.e., in the estimated one). Clearly, the smaller the
value of FN rate, the more desirable it is. Although there are two more common tree error measures (False Positive (FP rate) and Robinson-Foulds (RF) rate), all of them are identical when true and estimated trees are binary. In this thesis, we worked with binary trees only.

\subsection{Discordance between Gene and Species Phylogeny}\index{Gene tree discordance}
We now discuss the issue of discordance between a gene tree and species tree and the reasons behind it. A species tree represents the evolutionary history among a set of species via the process of speciation. On the contrary, a gene tree represents the evolution of a particular ``gene" within a group of species. When species are split by speciation, the gene copies within species are also split into separate lineages of descent.  Within each such lineage, the gene trees continue branching and descending through time. Thus, the gene trees are contained within the branches of the species trees~\cite{maddison1997gene}.
However, when gene copies are sampled from various species, the gene tree relating these copies might disagree with the species phylogeny. Fig.~\ref{fig:discordance} shows an example of discordance between a species tree and a gene tree. Here, species $B$ and $C$ are ``sister" species. However, in the gene lineage, $C$ is closer to $D$ than $B$. This discord can arise from the horizontal transfer, incomplete lineage sorting, and gene duplication and extinction~\cite{maddison1997gene}. 

\begin{figure}[!tb]
	\centering
	\includegraphics[width=0.33\textwidth]{Figure/discordance.eps}
	\caption{Gene tree-species tree discordance. A species tree (given in block diagram) and a gene tree (given in line diagram) on the same set $\{A,B,C,D\}$ of taxa with different topologies.}
	\label{fig:discordance}
\end{figure}



\subsubsection{Incomplete Lineage Sorting}\index{Incomplete lineage sorting (ILS)}

%Incomplete lineage sorting (ILS) refers to the failure of two gene lineages to coalesce at their speciation point. Also known as deep coalescence, this process is best understood under the multi-species coalescent (MSC) model~\cite{degnan2006discordance, degnan2005gene}. This model explains the evolutionary process by going backward in time and connecting gene lineages to a common ancestor through a process of ``coalescence" of lineage pairs. This is explained with an example in the Appendix.
%\section{Incomplete Lineage Sorting}%\index{Incomplete lineage sorting (ILS)}

Incomplete lineage sorting (ILS) refers to the failure of two gene lineages to coalesce at their speciation point. Also known as deep coalescence, this process is best understood under the coalescent model~\cite{degnan2006discordance, degnan2005gene}. This model explains evolutionary process by going backwards in time and connecting gene lineages to a common ancestor through a process of ``coalescence" of lineage pairs. In this model, each species is treated as a population of individuals, having a pair of alleles for each gene. The present day variants of a gene (known as alleles) are then traced back in time across successive generations by following the ancestral alleles in the previous generation from which this given alleles evolved. Eventually a point is reached where two alleles coalesce (i.e., they find a common ancestor). The multi-species coalescent (MSC) model is the extension of this general coalescent framework where multiple randomly mating populations corresponding to multiple species are present.
%, nei1987molecular, tajima1983evolutionary, takahata1989gene
\begin{figure}[!tb]
	\centering
	\includegraphics[width=0.33\textwidth]{Figure/ils.eps}
	\caption{Example of gene tree-species tree discordance due to incomplete lineage sorting. Going back in time, the gene copies within species $B$ and $C$ first meet at their corresponding speciation point, but fail to coalesce. Both the lineages (dashed and solid black lines) exist on deeper ancestral branch. The gene from $C$ first coalesces with the gene from species $D$, and subsequently with the gene from $B$.
	}
	\label{fig:ils}
\end{figure}
Under the MSC model, ILS can be a source of gene tree discordance, as the common ancestry of gene copies at a single locus can extend deeper than speciation events. The larger the effective population size and the shorter the branch length of the evolutionary tree, the greater the chance of ILS or deep coalescence to occur~\cite{maddison1997gene}.

Figure~\ref{fig:ils} shows an example of discordance due to ILS. The gene copies within species $B$ and $C$ first meet at their corresponding speciation point as we go back in time. The speciation point is the most recent common ancestor of species $B$ and $C$. However, the gene copies fail to coalesce here. Both of these copies go further back in time, resulting in two gene lineages on deeper ancestral branch. The extra lineage is shown by the dashed red lines in Figure~\ref{fig:ils}. Then the gene from $C$ first coalesces with the gene from species $D$, and subsequently with the gene from $B$.

\begin{comment}

\subsubsection{Quartet Score}
The quartet score~\cite{mirarab2014astral} measures the similarity between a candidate species tree $T$ and the input gene trees, and is computed as follows. Each input gene tree is decomposed into its induced set of quartet trees (i.e., unrooted trees formed by picking four leaves). The quartet support score of a given candidate species tree $T$ is the total, overall the input gene trees, of the number of induced quartet trees that $T$ agrees with. This score is maximized by ASTRAL to estimate the species tree. 

\subsubsection{Triplet Score}
The triplet score~\cite{islam2019stelar}, similar to quartet score, counts the number of induced triplet trees (i.e., rooted trees formed by picking three leaves) of the candidate species tree that agrees with the input gene trees. This criterion is maximized by STELAR. 

\subsubsection{Pseudo-likelihood}
It measures the likelihood of a candidate species tree given the triplet distribution of the input gene trees and maximized by MP-EST~\cite{mpest} as an estimator of the species tree.
\end{comment}

\subsubsection{Statistical consistency}\index{Statistical consistency}

A species tree reconstruction method is said to be statistically consistent under a particular model of evolution if the probability of returning the true species tree converges to one as the amount of data increases. Let the set of genes in a study be $\mathcal{G} = \{g_1, g_2, \dots, g_M\}$. Let $s_i$ be the number of sites in $g_i$ ($1 \leq i \leq M$). Then A species tree estimation method is statistically consistent if the estimated species tree converges to the true species tree as $M \rightarrow \infty$, $\underset{1 \leq i \leq M}{s_i} \rightarrow \infty$.


\subsubsection{Optimization criteria for species tree estimation}
Various criteria are statistically consistent estimators of the true species tree under the MSC model. Each of them measures how well a candidate species tree can summarize the input gene trees. Three examples are as follows:
\begin{itemize}
	\item Quartet score \cite{mirarab2014astral}: The quartet score of a given candidate species tree $T$ is the total, across all the input gene trees, of the number of induced quartet trees that $T$ agrees with. This score is maximized by ASTRAL to estimate the species tree.
	\item Triplet score \cite{islam2019stelar}: Similar to quartet score, it counts the number of induced triplet trees (i.e., rooted trees formed by picking three leaves) of the candidate species tree that agrees with the input gene trees.
	\item Pseudo-likelihood \cite{mpest}: It measures the likelihood of a candidate species tree given the triplet distribution of the input gene trees.
\end{itemize}


\section{Multiple Sequence Alignment}
Multiple sequence alignment (MSA) task seeks to arrange three or more nucleotide/protein sequences to infer homology, considering various biological phenomena (i.e., evolutionary history, 3D structure etc.). The output is a matrix in which the input nucleotide/protein sequences are the rows and each column (i.e., site) has characters which are homologous which means all those letters descend from the same letter of a common ancestor). The aligned sequences reflect historical substitution, insertion and deletion of genetic materials which are represented as gaps. Accurately recovering these properties through MSA is necessary to accomplish a biological objective such as inferring the evolutionary history relating the sequences known as phylogenetic trees. While computing MSAs, various computational methods and criteria are used to make hypotheses about homology. But the goal of MSA is entirely biological. Figure~\ref{fig:msa_io} illustrates this problem using an example where four hypothetical protein sequences are to be aligned by placing ``necessary'' gaps. %In this work, we focus on the MSA tasks in the context of phylogeny estimation which usually functions in two steps. Firstly, the given sequences are aligned using an MSA tool, thereafter a phylogenetic tree is inferred on obtained alignment. The goodness of estimated phylogenetic trees largely rely on the properties of the corresponding alignment. Thus, choosing the ``most appropriate'' MSA method  is important in the phylogenetic context.

\begin{figure}[!htbp]
	\centering
	\includegraphics[width=0.8\textwidth]{cybernatics/Figure/msa_io}
	\caption{A hypothetical instance of MSA task with four protein sequences.} 
	\label{fig:msa_io}
\end{figure}

\graphicspath{{cybernatics/supp/}}
\input{cybernatics/supp/supp_description} 
